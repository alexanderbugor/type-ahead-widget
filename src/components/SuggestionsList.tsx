import React from 'react';
import { WithStyles, withStyles, createStyles } from '@material-ui/core/styles';
import { default as MuiList } from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import { List } from '../list';

const styles = createStyles({
  root: {
    overflow: 'auto',
    width: '99.5%',
    height: '100%',
    maxHeight: '160px',
    border: '1px solid #2C397F',
    borderRadius: '4px',
  },
});

interface Props extends WithStyles<typeof styles> {
  list: List;
  onSuggestionSelect: (value: string) => void;
}

const SuggestionsList = ({ list, classes, onSuggestionSelect }: Props) => {
  if (!list.length) {
    return null;
  }
  
  return (
    <div className={classes.root}>
      <MuiList>
        {list.map((item, index) => (
          <ListItem button key={index} onClick={() => onSuggestionSelect(item)}>
            <ListItemText primary={item} />
          </ListItem>
        ))}
      </MuiList>
    </div>
  );
};

export default withStyles(styles)(SuggestionsList);
