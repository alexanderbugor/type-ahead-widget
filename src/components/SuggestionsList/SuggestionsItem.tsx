import React from 'react';
import { WithStyles, withStyles, createStyles } from '@material-ui/core/styles';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import { Suggestion } from './SuggestionsList';
import HighlightMatch from '../HighlightMatch/HighlightMatch';

const styles = createStyles({
  active: {
    backgroundColor: '#00000008',
  },
});

interface Props extends WithStyles<typeof styles> {
  item: Suggestion;
  index: number;
  activeItemIndex: number;
  onSuggestionSelect: (value: string) => void;
}

class SuggestionsItem extends React.PureComponent<Props> {
  ref = React.createRef<HTMLDivElement>();

  componentDidMount() {
    if (this.isItemActive()) {
      this.scrollItemIntoView();
    }
  }

  componentDidUpdate(prevProps: Props) {
    if (this.props.activeItemIndex !== prevProps.activeItemIndex && this.isItemActive()){
      this.scrollItemIntoView();
    }
  }

  scrollItemIntoView = () => {
    this.ref.current.scrollIntoView({
      behavior: 'smooth',
      block: 'start',
    });
  }

  isItemActive = () => {
    const { index, activeItemIndex } = this.props;

    return activeItemIndex === index;
  }
  
  render() {
    const {
      index,
      item,
      classes,
      onSuggestionSelect,
    } = this.props;

    return (
      <div ref={this.ref}>
        <ListItem
          className={this.isItemActive() ? classes.active : null}
          button
          onClick={() => onSuggestionSelect(item.item)}
        >
          <ListItemText
            primary={
              <HighlightMatch matches={item.matches}>
                {item.item}
              </HighlightMatch>
            }
          />
        </ListItem>
      </div>
    );
  }
}

export default withStyles(styles)(SuggestionsItem);
