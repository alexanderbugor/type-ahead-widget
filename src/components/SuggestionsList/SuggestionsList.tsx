import React from 'react';
import Fuse from 'fuse.js';
import { WithStyles, withStyles, createStyles } from '@material-ui/core/styles';
import { default as MuiList } from '@material-ui/core/List';
import SuggestionsItem from './SuggestionsItem';

const styles = createStyles({
  root: {
    overflow: 'auto',
    width: '99.5%',
    height: '100%',
    maxHeight: '160px',
    border: '1px solid #2C397F',
    borderRadius: '4px',
  },
  active: {
    backgroundColor: '#00000040',
  },
});

export type Suggestion = {
  item: string;
  matches: readonly Fuse.FuseResultMatch[];
  ref?: React.RefObject<HTMLDivElement>;
};

interface Props extends WithStyles<typeof styles> {
  list: Suggestion[];
  onSuggestionSelect: (value: string) => void;
}

type State = {
  activeItemIndex: number;
};

class SuggestionsList extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      activeItemIndex: 0,
    };
  }

  componentDidMount() {
    document.addEventListener('keydown', this.handleKeyDown, false);
  }

  handleKeyDown = (event: KeyboardEvent) => {
    const { activeItemIndex } = this.state;
    const { list, onSuggestionSelect } = this.props;

    if (event.keyCode === 38 && activeItemIndex > 0) {
      this.setState((prevState: State) => ({
        activeItemIndex: prevState.activeItemIndex - 1,
      }));
    }

    if (event.keyCode === 40 && activeItemIndex < list.length - 1) {
      this.setState((prevState: State) => ({
        activeItemIndex: prevState.activeItemIndex + 1,
      }));
    }

    if (event.keyCode === 13) {
      onSuggestionSelect(list[activeItemIndex].item);
    }
  };

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleKeyDown, false);
  }

  render() {
    const { list, classes, onSuggestionSelect } = this.props;
    const { activeItemIndex } = this.state;

    if (!list.length) {
      return null;
    }

    return (
      <div className={classes.root}>
        <MuiList>
          {list.map((item, index) => (
            <SuggestionsItem
              key={index}
              item={item}
              index={index}
              activeItemIndex={activeItemIndex}
              onSuggestionSelect={onSuggestionSelect}
            />
          ))}
        </MuiList>
      </div>
    );
  }
}

export default withStyles(styles)(SuggestionsList);
