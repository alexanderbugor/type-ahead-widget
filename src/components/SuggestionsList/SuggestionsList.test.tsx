import React from 'react';
import { shallow } from 'enzyme';
import Fuse from 'fuse.js';
import { SuggestionsList } from './SuggestionsList';

type MockListType = {
  item: string;
  matches: readonly Fuse.FuseResultMatch[];
};

describe('SuggestionsList', () => {
  const indices = ([1, 4] as unknown) as Fuse.RangeTuple[];
  const list: MockListType[] = [
    { item: 'Banana', matches: [{ indices, value: 'Banana' }] },
    { item: 'Mango', matches: [{ indices, value: 'Mango' }] },
    { item: 'Mangosteen', matches: [{ indices, value: 'Mangosteen' }] },
  ];

  const props = {
    classes: { root: 'MOCK_CLASS' },
    onSuggestionSelect: jest.fn(),
    list,
  };

  it('should not render an empty list', () => {
    const root = shallow(<SuggestionsList {...props} list={[]} />);

    expect(root.find('div')).not.toExist();
  });

  it('should render and match snapshot', () => {
    const root = shallow(<SuggestionsList {...props} />);

    expect(root).toMatchSnapshot();
  });

  describe('handleKeyDown', () => {
    const root = shallow<SuggestionsList>(<SuggestionsList {...props} />);

    root.setState({
      activeItemIndex: 1,
    });

    it('should handle up arrow key pressed', () => {
      root.instance().handleKeyDown({ keyCode: 38 } as KeyboardEvent);

      expect(root.instance().state.activeItemIndex).toEqual(0);
    });

    it('should handle down arrow key pressed', () => {
      root.instance().handleKeyDown({ keyCode: 40 } as KeyboardEvent);

      expect(root.instance().state.activeItemIndex).toEqual(1);
    });

    it('should handle Enter key pressed', () => {
      root.instance().handleKeyDown({ keyCode: 13 } as KeyboardEvent);

      expect(props.onSuggestionSelect).toHaveBeenCalledWith('Mango');
    });
  });
});
