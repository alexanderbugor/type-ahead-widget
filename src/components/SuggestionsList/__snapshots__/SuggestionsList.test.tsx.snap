// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestionsList should render and match snapshot 1`] = `
<div
  className="MOCK_CLASS"
>
  <WithStyles(ForwardRef(List))>
    <WithStyles(SuggestionsItem)
      activeItemIndex={0}
      index={0}
      item={
        Object {
          "item": "Banana",
          "matches": Array [
            Object {
              "indices": Array [
                1,
                4,
              ],
              "value": "Banana",
            },
          ],
        }
      }
      key="0"
      onSuggestionSelect={[MockFunction]}
    />
    <WithStyles(SuggestionsItem)
      activeItemIndex={0}
      index={1}
      item={
        Object {
          "item": "Mango",
          "matches": Array [
            Object {
              "indices": Array [
                1,
                4,
              ],
              "value": "Mango",
            },
          ],
        }
      }
      key="1"
      onSuggestionSelect={[MockFunction]}
    />
    <WithStyles(SuggestionsItem)
      activeItemIndex={0}
      index={2}
      item={
        Object {
          "item": "Mangosteen",
          "matches": Array [
            Object {
              "indices": Array [
                1,
                4,
              ],
              "value": "Mangosteen",
            },
          ],
        }
      }
      key="2"
      onSuggestionSelect={[MockFunction]}
    />
  </WithStyles(ForwardRef(List))>
</div>
`;
